cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -90,13 +93,82 @@
 import java.util.Queue;
 import java.util.Set;
 import java.util.concurrent.TimeUnit;
+import java.util.stream.IntStream;
 import javax.annotation.Nullable;
 
 /**
- * Service that receives program status notifications and persists to the store.
+ * Service that creates children services, each to handle a single partition of program status events topic
+ */
+public class ProgramNotificationSubscriberService extends AbstractIdleService {
+  private final MessagingService messagingService;
+  private final CConfiguration cConf;
+  private final MetricsCollectionService metricsCollectionService;
+  private final ProvisionerNotifier provisionerNotifier;
+  private final ProgramLifecycleService programLifecycleService;
+  private final ProvisioningService provisioningService;
+  private final ProgramStateWriter programStateWriter;
+  private final TransactionRunner transactionRunner;
+  private final Store store;
+  private final RunRecordMonitorService runRecordMonitorService;
+  private final Service delegate;
+
+  @Inject
+  public ProgramNotificationSubscriberService(MessagingService messagingService, CConfiguration cConf,
+                                       MetricsCollectionService metricsCollectionService,
+                                       ProvisionerNotifier provisionerNotifier,
+                                       ProgramLifecycleService programLifecycleService,
+                                       ProvisioningService provisioningService,
+                                       ProgramStateWriter programStateWriter, TransactionRunner transactionRunner,
+                                       Store store,
+                                       RunRecordMonitorService runRecordMonitorService) {
+
+    this.messagingService = messagingService;
+    this.cConf = cConf;
+    this.metricsCollectionService = metricsCollectionService;
+    this.provisionerNotifier = provisionerNotifier;
+    this.programLifecycleService = programLifecycleService;
+    this.provisioningService = provisioningService;
+    this.programStateWriter = programStateWriter;
+    this.transactionRunner = transactionRunner;
+    this.store = store;
+    this.runRecordMonitorService = runRecordMonitorService;
+    List<Service> children = new ArrayList<>();
+    String topicPrefix = cConf.get(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC);
+    int numPartitions = cConf.getInt(Constants.AppFabric.PROGRAM_STATUS_EVENT_NUM_PARTITIONS);
+    //Add bare one - we always listen to it
+    children.add(createChildService("program.status", topicPrefix));
+    //If number of partitions is more than 1 - create partitioned services
+    if (numPartitions > 1) {
+      IntStream.range(0, numPartitions).forEach(
+        i -> children.add(createChildService("program.status." + i, topicPrefix + i)));
```

Comment:
Do you have a subscribe that subscribe to the old topic too? It's for handling programs that starts before the upgrade. Maybe ut can simply be piggy backing on the subscriber for partition 0.

Reply:
Yes, it's line 139 right before the if